components/style/data.rs

```diff
@@ -8,24 +8,191 @@ use properties::ComputedValues;
+    fn is_previous(&self) -> bool {
+        use self::NodeDataStyles::*;
+        match *self {
+            EmptyPrevious | Previous(_) => true,
+            _ => false,
+        }
+    }
+}
+
+/// Transient data used by the restyle algorithm. This structure is instantiated
+/// either before or during restyle traversal, and is cleared at the end of node
+/// processing.
+#[derive(Debug)]
+pub struct RestyleData {
+    // FIXME(bholley): Start adding the fields from the algorithm doc.
+    pub _dummy: u64,
+}
+
+impl RestyleData {
+    fn new() -> Self {
+        RestyleData {
+            _dummy: 42,
         }
     }
 }
 
+/// Style system data associated with a node.
+///
+/// In Gecko, this hangs directly off a node, but is dropped when the frame takes
+/// ownership of the computed style data.
+///
+/// In Servo, this is embedded inside of layout data, which itself hangs directly
+/// off the node. Servo does not currently implement ownership transfer of the
+/// computed style data to the frame.
+///
+/// In both cases, it is wrapped inside an AtomicRefCell to ensure thread
+/// safety.
+#[derive(Debug)]
+pub struct NodeData {
+    styles: NodeDataStyles,
+    pub restyle: Option<RestyleData>,
+}
+
+impl NodeData {
+    pub fn new() -> Self {
+        NodeData {
+            styles: NodeDataStyles::Uninitialized,
+            restyle: None,
+        }
+    }
+
+    pub fn has_current_styles(&self) -> bool {
+        match self.styles {
+            NodeDataStyles::Current(_) => true,
+            _ => false,
+        }
+    }
+
+    pub fn get_current_styles(&self) -> Option<&NodeStyles> {
+        match self.styles {
+            NodeDataStyles::Current(ref s) => Some(s),
+            _ => None,
+        }
+    }
+
+    pub fn current_styles(&self) -> &NodeStyles {
+        self.get_current_styles().expect("Calling current_styles before or during styling")
+    }
+
+    // Servo does lazy pseudo computation in layout and needs mutable access
+    // to the current styles
+    #[cfg(not(feature = "gecko"))]
+    pub fn current_pseudos_mut(&mut self) -> &mut PseudoStyles {
+        match self.styles {
+            NodeDataStyles::Current(ref mut s) => &mut s.pseudos,
+            _ => panic!("Calling current_pseudos_mut before or during styling"),
+        }
+    }
+
+    pub fn previous_styles(&self) -> Option<&NodeStyles> {
+        match self.styles {
+            NodeDataStyles::EmptyPrevious => None,
+            NodeDataStyles::Previous(ref s) => Some(s),
+            _ => panic!("Calling previous_styles without having gathered it"),
+        }
+    }
+
+    pub fn previous_styles_mut(&mut self) -> Option<&mut NodeStyles> {
+        match self.styles {
+            NodeDataStyles::EmptyPrevious => None,
+            NodeDataStyles::Previous(ref mut s) => Some(s),
+            _ => panic!("Calling previous_styles without having gathered it"),
+        }
+    }
+
+    pub fn gather_previous_styles<F>(&mut self, f: F)
```

Comment:
I think to stick to the convention that `Option` has, we should call this `take_previous_styles`. I think the fact that the closure is only executed if it's `Uninitialized` is not documented nowhere.

Probably the more rusty name is something like `take_previous_styles_or_else`.

Also, if we change to `Option<NodeStyles>` instead of `EmptyPrevious`, this function would be way more idiomatic if you ask me.